pkg/monitortestlibrary/pathologicaleventlibrary/duplicated_events_special.go

```diff
@@ -60,6 +61,45 @@ func (s *singleEventThresholdCheck) Test(events monitorapi.Intervals) []*junitap
 	return []*junitapi.JUnitTestCase{success}
 }
 
+// NamespacedTest is just like Test() except it creates junits per namespace.
+func (s *singleEventThresholdCheck) NamespacedTest(events monitorapi.Intervals) []*junitapi.JUnitTestCase {
+	nsResults := map[string]*eventResult{}
+
+	for _, e := range events {
+		namespace := e.StructuredLocator.Keys[monitorapi.LocatorNamespaceKey]
+
+		// We only create junit for known namespaces
+		if !platformidentification.KnownNamespaces.Has(namespace) {
+			namespace = ""
+		}
+
+		if _, ok := nsResults[namespace]; !ok {
+			tmp := &eventResult{}
+			nsResults[namespace] = tmp
+		}
+
+		if s.matcher.Allows(e, "") {
+			msg := fmt.Sprintf("%s - %s", e.Locator, e.StructuredMessage.HumanMessage)
+			times := GetTimesAnEventHappened(e.StructuredMessage)
+			switch {
+			case s.failThreshold > 0 && times > s.failThreshold:
+				nsResults[namespace].failures = append(nsResults[namespace].failures, fmt.Sprintf("event [%s] happened %d times", msg, times))
+			case times > s.flakeThreshold:
+				nsResults[namespace].flakes = append(nsResults[namespace].flakes, fmt.Sprintf("event [%s] happened %d times", msg, times))
+			}
+		}
+	}
+
+	for ns, result := range nsResults {
```

Comment:
Wondering if you can do it without double processing the loop, initializing all namespaces then removing the empty ones

If you had a function like below could you call
nsResults[namespace] = addEventResult(nsResults[namespace], fmt.Sprintf("event [%s] happened %d times", msg, times), true)

etc.

func addEventResult(nsResult *eventResult{}, msg string, flake bool) *eventResult {
  if nsResult == nil {
     nsResult = &eventResult{}
  }

 if flake {
  nsResult.flakes = append(nsResult.flakes, msg)
} else {
  nsResult.failures = append(nsResult.failures, msg)
}
return nsResult

}

Reply:
That snippet won't work as is, you would probably have to pass in the map and the namespace and then do the check in the function.  Just doesn't feel right initializing it all and then removing the empty ones.. 

if _, ok := nsResults[namespace]; !ok {
			tmp := &eventResult{}
			nsResults[namespace] = tmp
		}

Reply:
[addEventResult_function.txt](https://github.com/openshift/origin/files/14058922/addEventResult_function.txt)

wonder if this patch would do it

Reply:
>> Just doesn't feel right initializing it all and then removing the empty ones.

Good instinct.

I pushed a commit that will track only namespaces with failures or flakes so that we only traverse the loop once.